apollo-router-core/src/layers/cache.rs

```diff
@@ -41,80 +46,92 @@ where
-            Some(Ok(value)) => Box::pin(futures::future::ready(Ok((self.response_fn)(
-                request, value,
-            )))),
-            Some(Err(err)) => Box::pin(futures::future::ready(Err(BoxError::from(err)))),
+        match key {
             None => {
-                let delegate = self.inner.call(request).err_into();
-                Box::pin(async move {
-                    let response = delegate.await;
-
-                    match &response {
-                        Ok(result) => {
-                            let value = value_fn(result);
-                            cache.insert(key, Ok(value));
-                        }
-                        Err(err) => {
-                            cache.insert(key, Err(err.to_string()));
-                        }
+                // The key function returned none, so the request was uncachable.
+                self.inner.call(request).err_into().boxed()
+            }
+            Some(key) => {
+                let value = self.cache.get(key);
+                match value {
+                    None => {
+                        // This is a new value. Create a sentinel and lock it.
+                        // The sentinel is an RWLock over an initially empty option.
+                        // Once the sentinel is created it is locked it so that readers must wait to read.
+                        // The sentinel is then put into the cache so that any subsequent read
+                        // will pick this up and have to wait.
+                        let sentinel = Arc::new(RwLock::new(None));
+                        let mut guard = futures::executor::block_on(sentinel.clone().write_owned());
+                        self.cache.insert(key.clone(), sentinel);
+
+                        let value_fn = self.value_fn;
+                        self.inner
+                            .call(request)
+                            .err_into()
+                            .then(move |response| async move {
+                                // We got the response now. Update the sentinel and release the guard.
+                                let sentinel_value = match &response {
+                                    Ok(response) => Ok((value_fn)(response).clone()),
+                                    Err(e) => Err(e.to_string()),
+                                };
+                                let _ = guard.insert(sentinel_value);
+                                response
+                            })
+                            .boxed()
                     }
-                    response
-                })
+                    Some(value) => {
+                        // The value has been populated, we need to await on the sentinel.
+                        // The read lock on the sentinel is released when it is populated allowing the
+                        // current read to proceed.
+                        let response_fn = self.response_fn;
```

Comment:
Is it possible for a sentinel to be evicted by the cache before the result is retrieved?

Reply:
Yes, it's definitely different behavior to before. 
It could be argued that users may find ttl easier to reason about, but it does come down to it being a tradeoff by using the sentinel approach.